In [13]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
plt.rcParams['figure.figsize']=(10,6)
from scipy.sparse import csr_matrix
from sklearn.model_selection import train_test_split
import scipy

### Reading data set with 50 features

In [3]:
df=pd.read_csv("C:\\Users\\athar\\Downloads\\ML_project\\dataset.csv")

In [4]:
# get ylabel
df_Y=df['y']
df_Y.head()

0     9
1     7
2     8
3    10
4    16
Name: y, dtype: int64

In [5]:
# get xlabel by deleting ylabel
del df['y']
df.head()

0  1  2  3  4  5  6  7  8  9  ...   1734  1735  1736  1737  1738  1739  \
0  0  0  0  1  0  0  0  0  0  0  ...      0     0     0     0     0     0   
1  0  0  0  0  0  0  0  0  0  0  ...      0     0     1     0     0     0   
2  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   
3  0  0  0  0  1  0  0  0  0  0  ...      0     0     0     0     0     0   
4  0  1  0  1  0  0  0  0  0  0  ...      0     0     0     0     0     0   

   1740  1741  1742  1743  
0     0     0     0     0  
1     0     0     0     1  
2     2     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 1744 columns]

In [6]:
df_X=df
df_X.head()

0  1  2  3  4  5  6  7  8  9  ...   1734  1735  1736  1737  1738  1739  \
0  0  0  0  1  0  0  0  0  0  0  ...      0     0     0     0     0     0   
1  0  0  0  0  0  0  0  0  0  0  ...      0     0     1     0     0     0   
2  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   
3  0  0  0  0  1  0  0  0  0  0  ...      0     0     0     0     0     0   
4  0  1  0  1  0  0  0  0  0  0  ...      0     0     0     0     0     0   

   1740  1741  1742  1743  
0     0     0     0     0  
1     0     0     0     1  
2     2     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 1744 columns]

In [7]:
# split the data into train data and test data
X_train, X_test, Y_train, Y_test = train_test_split(df_X, df_Y, test_size=0.33, random_state=42)

In [8]:
# converting into matrix for applicability in python
X_train_s=X_train.as_matrix()
X_test_s=X_test.as_matrix()

In [9]:
Y_train_m=Y_train.as_matrix()
Y_test_m=Y_test.as_matrix()
Y_train_m.shape

(13397,)

In [10]:
X_train_s.shape

(13397, 1744)

### import multinomial Naive Bayes classifier from sklearn library

In [36]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB
# declare a naive bayes classifier with alpha value 1 (Laplace smoothing)
# beta is n * alpha in this case where n is number of features
clf_pf = MultinomialNB(alpha=1)
clf_pf1 = MultinomialNB(alpha=1)

In [38]:
# this function does k-fold cross validation and returns average accuracy
def crossVal(X_train_s,Y_train_m):
    clf_pf = MultinomialNB(alpha=1)
    X_folds = np.array_split(X_train_s, 5)
    y_folds = np.array_split(np.ravel(Y_train_m), 5)
    scores = list()
    for k in range(5):
    # We use 'list' to copy, in order to 'pop' later on
        X_train = list(X_folds)
        X_test  = X_train.pop(k)
        X_train = np.concatenate(X_train)
        y_train = list(y_folds)
        y_test  = y_train.pop(k)
        y_train = np.concatenate(y_train)
        scores.append(clf_pf.fit(X_train, y_train).score(X_test, y_test))
    print(scores)
    mn2=scores
    z2=sum(mn2) / float(len(mn2))
    print("Cross-Validation Score of Naive Bayes %f" %z2)

### apply cross validation on train data and find accuracy on validation set

In [39]:
crossVal(X_train_s,Y_train_m)

[0.8436567164179104, 0.84291044776119406, 0.85628966032101528, 0.86151549085479662, 0.85703620754012688]
Cross-Validation Score of Naive Bayes 0.852282


### Check model accuracy on test data

In [40]:
clf_pf1.fit(X_train, Y_train).score(X_test, Y_test)

0.8562121212121212

### Inference: Training error of 14.77% and test accuracy of 85.62%. Let's increase the number of features and observe the accuracy for choosing the model

Tuning: Increasing number of features chosen from each category

In [3]:
# read data containing 100 features into a dataframe
# Creating dataset from a raw text file is done separately in ML_project_buildDataset
# This just saves repeated runs of the same code
df1=pd.read_csv("C:\\Users\\athar\\Downloads\\ML_project\\dataset_100f.csv")

In [4]:
# obtaining ylabel
df1_Y=df1['y']
df1_Y.head()

0     7
1     8
2    16
3    19
4    14
Name: y, dtype: int64

In [5]:
# obtaining xlabel by deleting ylabel
del df1['y']
df1.head()

Unnamed: 0  0  1  2  3  4  5  6  7  8  ...   3044  3045  3046  3047  3048  \
0           0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0   
1           1  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0   
2           2  0  1  0  0  0  0  0  0  2  ...      0     0     0     0     0   
3           3  0  0  0  0  1  0  0  0  0  ...      0     0     0     0     0   
4           4  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0   

   3049  3050  3051  3052  3053  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     0     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 3055 columns]

In [7]:
# removing unnecessary columns
df1.columns

Index(['Unnamed: 0', '0', '1', '2', '3', '4', '5', '6', '7', '8',
       ...
       '3044', '3045', '3046', '3047', '3048', '3049', '3050', '3051', '3052',
       '3053'],
      dtype='object', length=3055)

In [ ]:
del df1['Unnamed: 0']

In [10]:
df1_X=df1
df1_X.head()

0  1  2  3  4  5  6  7  8  9  ...   3044  3045  3046  3047  3048  3049  \
0  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   
1  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   
2  0  1  0  0  0  0  0  0  2  0  ...      0     0     0     0     0     0   
3  0  0  0  0  1  0  0  0  0  0  ...      0     0     0     0     0     0   
4  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   

   3050  3051  3052  3053  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 3054 columns]

In [12]:
# split the data into train data and test data
X_train1, X_test1, Y_train1, Y_test1 = train_test_split(df1_X, df1_Y, test_size=0.33, random_state=42)

In [13]:
# converting into matrix for applicability in python
X_train_s1=X_train1.as_matrix()
X_test_s1=X_test1.as_matrix()

In [14]:

Y_train_m1=Y_train1.as_matrix()
Y_test_m1=Y_test1.as_matrix()
Y_train_m1.shape

(13397,)

### apply cross validation on train data and find accuracy on validation set

In [41]:

crossVal(X_train_s1,Y_train_m1)

[0.86679104477611946, 0.86865671641791042, 0.87756625606569616, 0.87495334079880549, 0.86562150055991038]
Cross-Validation Score of Naive Bayes 0.870718


### Check model accuracy on test data


In [42]:
clf_pf.fit(X_train1, Y_train1).score(X_test1, Y_test1)

0.875

### Inference: Training error of 12.93% and test accuracy of 87.5%. We have achieved some improvement over the prvious case. Let's reduce the number of features and observe if adding features is worth it. As expected, error on train data is reduced.


In [14]:
 #read data containing 30 features into a dataframe
# Creating dataset from a raw text file is done separately in ML_project_buildDataset
# This just saves repeated runs of the same code
df2=pd.read_csv("C:\\Users\\athar\\Downloads\\ML_project\\dataset_30f.csv")
# obtain ylabel
df2_Y=df2['y']
df2_Y.head()

0    15
1    18
2    12
3     3
4     7
Name: y, dtype: int64

In [15]:
# obtain xlabel by removing ylabel
del df2['y']
del df2['Unnamed: 0'] # delete unnecessary columns
df2.head()

0  1  2  3  4  5  6  7  8  9  ...   1084  1085  1086  1087  1088  1089  \
0  0  0  0  0  0  0  0  1  0  0  ...      0     0     0     0     0     0   
1  0  0  0  0  0  0  0  1  0  1  ...      0     0     0     0     0     0   
2  0  1  0  0  0  0  1  0  0  0  ...      0     0     0     0     0     0   
3  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   
4  2  1  1  0  0  0  1  0  1  0  ...      0     0     0     0     0     1   

   1090  1091  1092  1093  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 1094 columns]

In [16]:
df2_X=df2
df2_X.head()

0  1  2  3  4  5  6  7  8  9  ...   1084  1085  1086  1087  1088  1089  \
0  0  0  0  0  0  0  0  1  0  0  ...      0     0     0     0     0     0   
1  0  0  0  0  0  0  0  1  0  1  ...      0     0     0     0     0     0   
2  0  1  0  0  0  0  1  0  0  0  ...      0     0     0     0     0     0   
3  0  0  0  0  0  0  0  0  0  0  ...      0     0     0     0     0     0   
4  2  1  1  0  0  0  1  0  1  0  ...      0     0     0     0     0     1   

   1090  1091  1092  1093  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 1094 columns]

### split the data into train data and test data


In [17]:
X_train2, X_test2, Y_train2, Y_test2 = train_test_split(df2_X, df2_Y, test_size=0.33, random_state=42)

In [18]:
# converting into matrix for applicability in python
X_train_s2=X_train2.as_matrix()
X_test_s2=X_test2.as_matrix()

In [19]:
Y_train_m2=Y_train2.as_matrix()
Y_test_m2=Y_test2.as_matrix()
Y_train_m2.shape

(13397,)

### apply cross validation on train data and find accuracy on validation set

In [61]:
crossVal(X_train_s2,Y_train_m2)

[0.83320895522388061, 0.84514925373134331, 0.84733109369167603, 0.84807764091078763, 0.84957073534901084]
Cross-Validation Score of Naive Bayes 0.844668


### Check model accuracy on test data

In [62]:
clf_pf.fit(X_train2, Y_train2).score(X_test2, Y_test2)

0.85030303030303034

### Inference: Training error of 15.54% and test accuracy of 85.03%. As expected, error on training data increased. By adding more features (3000 instead of 1000) we improve prediction accuracy by approximately 2.5%. This completes comparison for Naive Bayes classifier. Let's move on to SVM. 

In [20]:
# import SVM classifier
from sklearn.svm import SVC
from sklearn import svm

### cost factor 1 and linear kernel on data containing 50 features from each category

In [48]:
svc = svm.SVC(C=1, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train, Y_train).score(X_test, Y_test))

Accuracy of SVM : 0.873182


### cost factor 2 and linear kernel on data containing 30 features from each category

In [63]:
svc = svm.SVC(C=2, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train2, Y_train2).score(X_test2, Y_test2))

Accuracy of SVM : 0.855909


### cost factor 1 and linear kernel on data containing 30 features from each category

In [64]:
svc = svm.SVC(C=1, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train2, Y_train2).score(X_test2, Y_test2))

Accuracy of SVM : 0.861818


### improved accuracy on reducing the cost factor. Let's try to reduce it further

In [21]:
svc = svm.SVC(C=0.5, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train2, Y_train2).score(X_test2, Y_test2))

Accuracy of SVM : 0.866061


### Improved accuracy on further reducing cost factor. Let's try to reduce it even further

In [23]:
svc = svm.SVC(C=0.1, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train2, Y_train2).score(X_test2, Y_test2))

Accuracy of SVM : 0.883182


### Improved accuracy. Try C = 0.01

In [24]:
svc = svm.SVC(C=0.01, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train2, Y_train2).score(X_test2, Y_test2))

Accuracy of SVM : 0.882576


### Reduced accuracy. Therefore, C=0.1 is the appropriate cost factor

### cost factor 2 and linear kernel on data containing 50 features from each category


In [65]:
svc = svm.SVC(C=2, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train, Y_train).score(X_test, Y_test))

Accuracy of SVM : 0.872121


### Even here reducing C improves accuracy. Let's try to reduce it further

In [25]:
svc = svm.SVC(C=0.5, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train, Y_train).score(X_test, Y_test))

Accuracy of SVM : 0.876061


### As accuracy has improved, we will reduce C

In [26]:
svc = svm.SVC(C=0.1, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train, Y_train).score(X_test, Y_test))

Accuracy of SVM : 0.888939


### try reducing C even further

In [27]:
svc = svm.SVC(C=0.01, kernel='linear')
print("Accuracy of SVM : %f"%svc.fit(X_train, Y_train).score(X_test, Y_test))

Accuracy of SVM : 0.885909


### Accuracy reduced. Therefore, C = 0.1 is the appropriate point